<a href="https://colab.research.google.com/github/KunChen-tw/DeepLearning/blob/main/Practice_1_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Practice 1-4
if $y=x^2+1$ and $z=y^5$, find gradient $\left. \frac{dz}{dx} \right|_{x=1}$

In [ ]:
import torch
def dfunc(x):
  x = torch.tensor(float(x), requires_grad=True)
  y = x**2+1
  z = y**5
  z.backward()
  return x.grad

print(dfunc(1))


tensor(160.)


# 手算過程

跟據連鎖律
$$\frac{dz}{dx} = \frac{dz}{dy} \cdot \frac{dy}{dx}$$

首先，對 $y=x^2+1$ 求導數：
$$\frac{dy}{dx} = 2x$$

對 $z=y^5$ 求導數：
$$\frac{dz}{dy} = 5y^4$$

因此：
$$\frac{dz}{dx} = 5y^4 \cdot 2x = 10xy^4 $$

當 $x = 1$ 時
$$y = 1^2 + 1 = 2$$

將 $x=1$ , $ y=2$ 代入

$$\frac{dz}{dx} = 10 \cdot 1 \cdot 2^4 = 10 \cdot 16 = 160$$